### PR TITLE
APIM: Delete subscriptions when deleting products

### LINKED ISSUE #3408: APIM: Product should delete any Subscriptions it owns when it deletes
When you create an APIM Product it creates a Subscription. When you come to delete the Product, you will get an error saying you must delete the Subscription first

The ApiManagementDeleteProducts.json shows that an extra parameter is passed in the body 
`"deleteSubscriptions": true`

Map that field into the Code Generation

### PR DESCRIPTION
**Pull request by ross-p-smith**
This pull request fixes two APIM issues

# Goal Seeker behaviour
My previous work on APIM left a setting of `Recover: true` when creating the apim instance. This will cause an error if you were to delete the recording and not have an apim instance in a deleted state.

It would be better not to have this setting defined and allow it to error if there was a deleted APIM in your resource group - which is a smaller probability!

This would eventually be handled by having a goal seeking APIM creation #3433.

# Test deletion of products
When we create an APIM Product, it creates a Subscription for that Product. This causes problems when you delete the Product; however there is a flag to pass deleteSubscriptions to true which will fix this. Override the default delete behaviour and use the go sdk to carry out the deletion. This will close #3408

**How does this PR make you feel**:
gif

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains tests

### CODE REVIEW

File: v2/api/apimanagement/customizations/product_extensions.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+		parentName,
+		productName,
+		"*",
+		&armapimanagement.ProductClientDeleteOptions{DeleteSubscriptions: to.Ptr(true)})
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to delete product %q", productName)
+	}
+
+	// Is it safe to proceed as we have already deleted it?
+	return next(ctx, log, resolver, armClient, obj)
```
Comment (theunrepentantgeek) [inline v2/api/apimanagement/customizations/product_extensions.go:72]: Yes, this is safe. ASO will attempt the deletion, then update all necessary bookkeeping.

File: v2/api/apimanagement/customizations/product_extensions.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	parentName := id.Parent.Name
+
+	// Using armClient.ClientOptions() here ensures we share the same HTTP connection, so this is not opening a new
+	// connection each time through
+	clientFactory, err := armapimanagement.NewClientFactory(id.SubscriptionID, armClient.Creds(), armClient.ClientOptions())
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to create new apimClient")
+	}
+
+	_, err = clientFactory.NewProductClient().Delete(
```
Comment (theunrepentantgeek) [inline v2/api/apimanagement/customizations/product_extensions.go:60]: Is `Delete` synchronous? That is, finished by the time the call returns? Or does it return a long running operation you should monitor until the delete is complete?

**Review by theunrepentantgeek — CHANGES_REQUESTED** [review]
Only needs a couple of minor updates, then will be ready to merge.

File: v2/api/apimanagement/customizations/product_extensions.go (hunk 2 above)
Comment (ross-p-smith) [reply v2/api/apimanagement/customizations/product_extensions.go:60]: My understanding is that Delete is synchronous. According to https://learn.microsoft.com/en-us/azure/developer/go/management-libraries#long-running-operations they all begin with `Begin`.

Documentation of creating the actual APIM Service is a long running operation and is called out as one e.g. https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/apimanagement/armapimanagement#ServiceClient.BeginCreateOrUpdate

**Review by theunrepentantgeek — APPROVED** [review]

**Pull request merged**